File: BaekJoon/solved/C++/5464 주차장.cpp
#include <stdio.h>

int N, M;
int queue[2010]; // 주차할 차량
int park[105][2];//주차 공간/요금
int wei[2010]; //차량 무게
int front, rear;

void push(int data)
{
	rear++;
	rear %= 2010;
	queue[rear] = data;
}

int pop()
{
	front++;
	front %= 2010;
	return queue[front];
}

void BOJ_5464()
{
    scanf("%d %d", &N, &M);
	for (int i = 0; i < N; i++)
		scanf("%d", &park[i][1]); //무게당 요금
	for (int i = 1; i <= M; i++)
		scanf("%d", &wei[i]); //차량 당 무게

	long long int ans = 0; //총 수입
	M *= 2;
	for (int i = 0; i < M; i++)
	{
		int a;
		scanf("%d", &a);
		int lot = 0;
		if (a < 0) //주차장 나감
		{
			a *= -1;
			while (park[lot][0] != a) lot++; //나갈 차량
			park[lot][0] = 0;
			if (front != rear) //큐에 차량이 있으면 주차시키기
			{
				park[lot][0] = pop();
				ans += park[lot][1] * wei[park[lot][0]];
			}
		}
		else //주차장 들어옴
		{
			push(a);
			while (park[lot][0] != 0 && lot < N) lot++; //빈공간 찾기
			if(lot < N) //주차공간 있음
			{
				park[lot][0] = pop();
				ans += park[lot][1] * wei[park[lot][0]];
			}
		}
	}

	printf("%lld\n", ans);
}

int main()
{
    BOJ_5464();
    return (0);
}

File: BaekJoon/solved/JS/5464 주차장.js
let I = require("fs").readFileSync("./dev/stdin").toString().split("\n");

num = I[0].split(" ")
const N = +num[0];
const M = +num[1];

I = I.slice(1).map(e => +e);

let Q = [0], front = 0, rear = 0;

const push = (data) => {
    rear++;
    Q[rear] = data;
}

const pop = () => {
	front++;
	return Q[front];
}

const BOJ_5464 = () => {
    let W = I.slice(N, N + M); 
    let car = I.slice(N + M);
    let ans = 0;
    let park = I.slice(0, N).map(e => [0, e]);

    for (let i = 0; i < M*2; i++)
    {
        let lot = 0;
        if (car[i] < 0) {
            while (lot < N && park[lot][0] != -car[i]) 
                lot++;
            park[lot][0] = 0;
            if (front != rear)
            {
                park[lot][0] = pop();
                ans += park[lot][1] * W[park[lot][0] - 1];
            }
        }
        else {
            push(car[i]);
            while (lot < N && park[lot][0]) 
                lot++;
            if (lot < N)
            {
                park[lot][0] = pop();
                ans += park[lot][1] * W[park[lot][0] - 1];
            }
        }
    }
    console.log(ans);
}

BOJ_5464();